Mathlib/Analysis/Normed/Ring/Ultra.lean

```diff
@@ -0,0 +1,71 @@
+/-
+Copyright (c) 2024 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Analysis.Normed.Field.Basic
+import Mathlib.Analysis.Normed.Group.Ultra
+
+/-!
+# Ultrametric norms on rings with a multiplicative norm
+
+This file contains results on the behavior of norms in ultrametric normed rings.
+The norm must be multiplicative, not just submultiplicative.
+
+## Main results
+
+* `norm_intCast_le_one`:
+  the norm of the image of an integer in the division ring is always less than or equal to one
+
+## Implementation details
+
+A `[NormedRing R]` only assumes a submultiplicative norm. The weakest ring-like structure
+that has a bundled multiplicative norm is `[NormedDivisionRing K]`.
+Since the statements below hold in a multiplicative norm context, we can state them
+in an unbundled fashion using `[NormOneClass R]`.
```

Comment:
I don't agree with the statements in this docstring. The lemmas all hold _without_ assuming any multiplicativity, or even submultiplicativity, of the norm; in fact they don't use the existence of multiplication on R at all!

In fact one can actually prove all these lemmas only assuming `{R : Type*} [SeminormedAddGroup R] [One R] [NormOneClass R] [IsUltrametricDist R]`. But one has to give the typeclass machinery a little help in order to get it to recognise that there is a coercion from N or Z to R. I'm not sure this extra generality is worth the bother, so I guess it's easiest to use the weakest pre-existing typeclass that implies both `[SeminormedAddGroup R]` and `[AddGroupWithOne R]`. But division structure is certainly irrelevant.

Reply:
Fair, I'll modify the contents here to discuss NormOneClass without mentioning multiplicativity. However, it's my understanding that one doesn't really get NormOneClass on its own that often -- that instead, it occurs on rings that _do_ have multiplicative norms, like the examples discussed (Z_p). So I wanted to give a connection to that and explain why NormedRing isn't appropriate here, even though it has Norm and Ring in the name.

Reply:
I guess the important counterexample for my assertion about multiplicativity and NormOneClass are CLMs.